# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/x86/Lowering-x86.cpp
# Commit: aad37f6cbf03
# Full Hash: aad37f6cbf034f5ec3e0b04989aa025c7c899697
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-09-11 15:10:42
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 6.2: Establish dependency from tls for x86 callWithABI div/mod i64. r=lth
#   
#   x86 has few register so to do div/mod for i64 it call the runtime
#   and clobber almost all gp registers including WasmTlsReg.
#   To be able to call c++ runtime via Builtin thunk we need to set up WasmTlsReg.
# ==============================================================================

diff -r 05007ff436a5 -r aad37f6cbf03 js/src/jit/x86/Lowering-x86.cpp
--- a/js/src/jit/x86/Lowering-x86.cpp	Fri Sep 11 07:15:21 2020 +0000
+++ b/js/src/jit/x86/Lowering-x86.cpp	Fri Sep 11 08:05:01 2020 +0000
@@ -586,41 +586,66 @@
 }
 
 void LIRGeneratorX86::lowerDivI64(MDiv* div) {
+  MOZ_CRASH("We use MWasmBuiltinModI64 instead.");
+}
+
+void LIRGeneratorX86::lowerWasmBuiltinDivI64(MWasmBuiltinDivI64* div) {
+  MOZ_ASSERT(div->lhs()->type() == div->rhs()->type());
+  MOZ_ASSERT(IsNumberType(div->type()));
+
+  MOZ_ASSERT(div->type() == MIRType::Int64);
+
   if (div->isUnsigned()) {
-    lowerUDivI64(div);
+    LUDivOrModI64* lir = new (alloc())
+        LUDivOrModI64(useInt64FixedAtStart(div->lhs(), Register64(eax, ebx)),
+                      useInt64FixedAtStart(div->rhs(), Register64(ecx, edx)),
+                      useFixedAtStart(div->tls(), WasmTlsReg));
+    defineReturn(lir, div);
     return;
   }
 
-  LDivOrModI64* lir = new (alloc()) LDivOrModI64(
-      useInt64FixedAtStart(div->lhs(), Register64(eax, ebx)),
-      useInt64FixedAtStart(div->rhs(), Register64(ecx, edx)), tempFixed(esi));
+  LDivOrModI64* lir = new (alloc())
+      LDivOrModI64(useInt64FixedAtStart(div->lhs(), Register64(eax, ebx)),
+                   useInt64FixedAtStart(div->rhs(), Register64(ecx, edx)),
+                   useFixedAtStart(div->tls(), WasmTlsReg));
   defineReturn(lir, div);
 }
 
 void LIRGeneratorX86::lowerModI64(MMod* mod) {
+  MOZ_CRASH("We use MWasmBuiltinModI64 instead.");
+}
+
+void LIRGeneratorX86::lowerWasmBuiltinModI64(MWasmBuiltinModI64* mod) {
+  MDefinition* lhs = mod->lhs();
+  MDefinition* rhs = mod->rhs();
+  MOZ_ASSERT(lhs->type() == rhs->type());
+  MOZ_ASSERT(IsNumberType(mod->type()));
+
+  MOZ_ASSERT(mod->type() == MIRType::Int64);
+  MOZ_ASSERT(mod->type() == MIRType::Int64);
+
   if (mod->isUnsigned()) {
-    lowerUModI64(mod);
+    LUDivOrModI64* lir = new (alloc())
+        LUDivOrModI64(useInt64FixedAtStart(mod->lhs(), Register64(eax, ebx)),
+                      useInt64FixedAtStart(mod->rhs(), Register64(ecx, edx)),
+                      useFixedAtStart(mod->tls(), WasmTlsReg));
+    defineReturn(lir, mod);
     return;
   }
 
-  LDivOrModI64* lir = new (alloc()) LDivOrModI64(
-      useInt64FixedAtStart(mod->lhs(), Register64(eax, ebx)),
-      useInt64FixedAtStart(mod->rhs(), Register64(ecx, edx)), tempFixed(esi));
+  LDivOrModI64* lir = new (alloc())
+      LDivOrModI64(useInt64FixedAtStart(mod->lhs(), Register64(eax, ebx)),
+                   useInt64FixedAtStart(mod->rhs(), Register64(ecx, edx)),
+                   useFixedAtStart(mod->tls(), WasmTlsReg));
   defineReturn(lir, mod);
 }
 
 void LIRGeneratorX86::lowerUDivI64(MDiv* div) {
-  LUDivOrModI64* lir = new (alloc()) LUDivOrModI64(
-      useInt64FixedAtStart(div->lhs(), Register64(eax, ebx)),
-      useInt64FixedAtStart(div->rhs(), Register64(ecx, edx)), tempFixed(esi));
-  defineReturn(lir, div);
+  MOZ_CRASH("We use MWasmBuiltinDivI64 instead.");
 }
 
 void LIRGeneratorX86::lowerUModI64(MMod* mod) {
-  LUDivOrModI64* lir = new (alloc()) LUDivOrModI64(
-      useInt64FixedAtStart(mod->lhs(), Register64(eax, ebx)),
-      useInt64FixedAtStart(mod->rhs(), Register64(ecx, edx)), tempFixed(esi));
-  defineReturn(lir, mod);
+  MOZ_CRASH("We use MWasmBuiltinModI64 instead.");
 }
 
 void LIRGenerator::visitSubstr(MSubstr* ins) {